Add example of parsing array of sets to nested objects to cheatsheet

teq0 opened the pull request:
Signed-off-by: teq0 <craig.hooper@gmail.com>

This, or something like it, has been asked on Slack a number of times, I thought it might be a good candidate for the cheatsheet.

teq0 commented:
OK, need some help here. Build is failing because
```
8:11:50 PM: 	rules/parse_array_to_nested_objs evaluation failed for an unexpected, possibly internal reason (check for additional logs): couldn't find package declaration in module
```
But none of the other examples have package declarations. Is this because of the `output` section?

patrick-east commented:
Yea, if there is an output section then the docs will evaluate it and ensure that the right type of output is provided (there are helpers for like "expect error" situations to avoid us publishing broken examples).

The `module` section is expected to be a valid Rego module, including a package definition. If you omit the output it will just be syntax highlighted and not do any evaluations.

patrick-east commented on docs/content/policy-cheatsheet.md:
````diff
@@ -268,6 +268,98 @@ z := {"a": true, "b": "foo", "c": 4}
+This is an example of parsing an array of sets that have a known structure into nested objects.
+
+```live:rules/parse_array_to_nested_objs:query:read_only
+entity_array = [
````
I like the idea, but IMO we should maybe minimize the example a little bit more to try and keep the cheatsheet examples short and to the point (notice that this one is significantly larger than the others in here). 

Maybe we just do like a couple array entries, shorter names (`x`, `foo`, etc) and that will also simplify the comprehensions.

teq0 replied:
OK, will do.

patrick-east commented on docs/content/policy-cheatsheet.md:
```diff
@@ -268,6 +268,98 @@ z := {"a": true, "b": "foo", "c": 4}
+      }
+    }
+  }
+}
```
It might be good to add some inline comments just walking through the like logic of what this is doing so people can not only copy and re-use but (hopefully) understand and iterate/adapt it for their use-case.

teq0 replied:
I've condensed the query and added comments to the function, do they look OK?

patrick-east replied:
I like comments, only nit would be do move them to the side so its like policy on the left and comments on the right like:

```
do_the_thing() {		# Helper to do the thing
      a + b = 3			# That makes sure it equals 3 for some reasons 
      another_thing		# Finished up with another thing because X
}
```

teq0 replied:
Good idea. It looked a bit busy to me, but adding blank lines didn't really help.

teq0 commented:
Sorry @patrick-east, I've condensed the query and added comments to the function, but I can't figure out why the Netlify build is failing. Any clues? 

```
7:16:03 PM: Error: rules/parse_array_to_nested_objs evaluation failed for an unexpected, possibly internal reason (check for additional logs): invalid response while trying to get details about an evaluation failure
7:16:03 PM:     at /opt/build/repo/docs/website/scripts/live-blocks/src/preprocess/index.js:213:13
7:16:03 PM:     at process._tickCallback (internal/process/next_tick.js:68:7)
```

Also, do the comments look OK? Helpful? Need more whitespace or anything?

patrick-east commented:
Sorry for the trouble with the doc build.. Its not the most simple thing in the world. The error on the latest one is:
```
12:16:03 AM: --- Errors for ../../public/docs/edge/policy-cheatsheet/index.html ---
12:16:03 AM: Error: rules/parse_array_to_nested_objs evaluation failed for an unexpected, possibly internal reason (check for additional logs): invalid response while trying to get details about an evaluation failure
12:16:03 AM:     at /opt/build/repo/docs/website/scripts/live-blocks/src/preprocess/index.js:213:13
12:16:03 AM:     at process._tickCallback (internal/process/next_tick.js:68:7)
12:16:03 AM: (caused by) Error: invalid response while trying to get details about an evaluation failure
12:16:03 AM:     at localEval (/opt/build/repo/docs/website/scripts/live-blocks/src/preprocess/localEval.js:54:21)
12:16:03 AM:     at process._tickCallback (internal/process/next_tick.js:68:7)
12:16:03 AM: (caused by) Error: an opa error should either be the value undefined (indicating an undefined decision) or have a code
12:16:03 AM:     at new OPAErrors (/opt/build/repo/docs/website/scripts/live-blocks/src/errors.js:28:15)
12:16:03 AM:     at localEval (/opt/build/repo/docs/website/scripts/live-blocks/src/preprocess/localEval.js:49:29)
12:16:03 AM:     at process._tickCallback (internal/process/next_tick.js:68:7)
12:16:03 AM: It looks like there were some errors. Here's an overview:
12:16:03 AM: ../../public/docs/edge/policy-cheatsheet/index.html:
12:16:03 AM: 	rules/parse_array_to_nested_objs evaluation failed for an unexpected, possibly internal reason (check for additional logs): invalid response while trying to get details about an evaluation failure
12:16:03 AM: Makefile:68: recipe for target 'live-blocks-inject' failed
```
They important bit being 

```
12:16:03 AM: (caused by) Error: invalid response while trying to get details about an evaluation failure
12:16:03 AM:     at localEval (/opt/build/repo/docs/website/scripts/live-blocks/src/preprocess/localEval.js:54:21)
12:16:03 AM:     at process._tickCallback (internal/process/next_tick.js:68:7)
12:16:03 AM: (caused by) Error: an opa error should either be the value undefined (indicating an undefined decision) or have a code
```
Which is saying that it tried to parse the evaluation result, found an error, and wasn't instructed that the snippet is supposed to have an error (via the like `expect_*` tags https://github.com/open-policy-agent/opa/tree/master/docs#tags )

So what error did it have? Well, lets find out.. 

I added in some trusty `console.log()` calls into `docs/website/scripts/live-blocks/src/preprocess/localEval.js`.. cus helpful error output isn't there normally (I guess)
```
{16:09} ✘ ~/p/g/s/g/o/o/d/w/s/live-blocks:pr/2168 ◈ ❯ npm run preprocess ../../public/docs/edge/policy-cheatsheet/index.html

> live-blocks@0.0.0 preprocess /Users/patrick/projects/go/src/github.com/open-policy-agent/opa/docs/website/scripts/live-blocks
> node -r esm src/preprocess/index.js "../../public/docs/edge/policy-cheatsheet/index.html"

Processing ../../public/docs/edge/policy-cheatsheet/index.html...
Locating OPA version edge...
Found OPA version edge, available at /Users/patrick/projects/go/src/github.com/open-policy-agent/opa/opa_darwin_amd64
RUNNING: opa eval --fail --format=pretty -d /var/folders/9q/56f22r0x04v8506j6p43xw_80000gn/T/tmp-2423863UPrrG4o26D.rego --package parse_array_to_nested_objs some_array = [
  [ "product1", "type1", "id1", "view", { "name": "Audrey" } ],
  [ "product1", "type1", "id2", "write", { "name": "Brian" } ],
  [ "product1", "type2", "id3", "write", { "name": "Cynthia" } ],
  [ "product2", "type2", "id4", "view", { "name": "Derek" } ]
]

nestedobjs = objs {
  objs := parse_to_nestedobjs(some_array)
}
ERROR STDOUT : {
  "errors": [
    {
      "message": "expected body but got *ast.Rule"
    }
  ]
}


--- Errors for ../../public/docs/edge/policy-cheatsheet/index.html ---
ChainedError: rules/parse_array_to_nested_objs evaluation failed for an unexpected, possibly internal reason (check for additional logs): invalid response while trying to get details about an evaluation failure
```

So I guess the problem is that the query has a rule in it, which it doesn't like. Try removing `nestedobjs` and just calling `parse_to_nestedobjs(some_array)` in the query.

I think too I would maybe. drop the depth down a couple notches and remove the `product`, `id`, `type`, etc. IMO just like:

```rego
array = [
  [ "a", "x", {"foo1" 1} ],
  [ "a", "x", {"foo2": 2} ],
  [ "a", "y",  {"foo3": 3}],
]
```
So that we can maybe make it a little bit more generic, like `k0`, `k1`, etc for the keys at some level or something. As-is my concern is that its kind of a complex example and is going to be hard for someone to translate. The pattern though is just how to nest comprehensions and build objects from arrays so I'd prefer the focus on that.. if that makes sense.

teq0 commented:
I simplified it and moved the comments, I don't want to simplify it too much or it starts to just be obvious, but I think it's about the right level now.

Build is passing, but when I look at the netlify preview the new section isn't there, not sure what that means.

samgurtman commented:
It seems like the time complexity on this method is quite high. While this works for small datasets it becomes unfeasible with larger datasets. I'm wondering if a built in needs to be created.

teq0 commented:
> It seems like the time complexity on this method is quite high.
Good call. Is there a better algorithm using rego right now? The performance guidelines do say not to use arrays if you can help it, I'm assuming this kind of pattern isn't going to be used in a low latency situation.

I'm not sure what the signature would look like for a built-in, the input array is unstructured, the code needs to know what fields are keys for what level etc. Although I guess you could assume that they are in order and just give the depth, which treats the first n columns as keys, with everything below that being the objects.

samgurtman commented:
> I'm not sure what the signature would look like for a built-in, the input array is unstructured, the code needs to know what fields are keys for what level etc. Although I guess you could assume that they are in order and just give the depth, which treats the first n columns as keys, with everything below that being the objects.

My original thought was expanding `walk(x, [path, value])` so that x could be an output rather than just an input. Otherwise I think just a groupBy in the form of `groupBy(arr, idx)` could work. 

`groupBy([["123","read"],["123","write"],["456","write"]], 0)`

Output:
`[["123",["read","write"]], ["456", ["write"]]]`

Alternatively maybe an object would be a better ouput:
`{ "123": ["read", "write"], "456": ["write"]}`

patrick-east commented:
I like the idea of having a builtin function to basically act as the opposite of `walk`. The suggestions above sound pretty nice, we might want to open a new feature request on GitHub to track that and we can flesh out the design a little bit more.

Thinking more on this maybe we should hold off adding it to the cheatsheet/patterns section of the docs for now. With the performance issue it doesn't really seem like something we should be recommending policy authors should use.

teq0 commented:
Closing this as it may be replaced with a native function and due to concerns over possible performance issues with large datasets.
